### PR TITLE
[imagecache] Don't check special generated images for file changes every 24 hours

### PR DESCRIPTION
**Pull request by rmrector**
## Description
Don't check special generated images for file changes every 24 hours.

Special generated images like video thumbnails and picture folder thumbnails, as generated from the interface introduced in #23197 . 

Also includes a small update to class docs.

## Motivation and context
Special images didn't do this before the new interface, I didn't expect to get this "functionality" with the new interface, and it is a bit frustrating even for the basic images it should apply to right now so we don't need to proliferate it. My upcoming chapter thumbnail change will likely end up with a path that can't be 'stat'ed which fails here unnecessarily, and @ksooo may use the same for PVR.

This does technically also change the behavior of HTTP images and embedded music thumbs, which previously generated and saved the file hash when _first_ loaded, but I think the change is a good thing for HTTP and shouldn't affect outward behavior.

This is quick and a bit messy to avoid the cache behavior - once all the special loaded images are implemented we'll be in a better place to re-arrange and simplify this and other lists of "special image types" and `DecodeImageURL`.

## How has this been tested?
Quickly on Windows.

## What is the effect on users?
Keep classic behavior before the new interface goes out in a wide release.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by KarellenX** [comment]
Any chance of extending this for all images?

A common complaint is hard drives always spinning up for the hash check of images while scrolling through the library. When this happens, the GUI freezes until the drives have spun up and the check completed. This can be quite annoying.

The workaround I advise for advanced users is to edit the LastCheckedDate in the DB to a year into the future.

I have never really understood why images need to be checked every 24 hours, when we never check for anything else... like the more important video file.

Maybe a setting to nominate when the check occurs - Never, 24hrs, 48hrs, weekly, monthly?

**Comment by rmrector** [comment]
We need it for regular images because we would have even more complaints about replaced image files that never update in Kodi. We already DO get complaints about that, but currently they are easily answered with "wait 24 hours."

The current behavior is silly, yes, but just adding a knob to twiddle won't help - folks will switch it on/off/highest/lowest based on their desire to fiddle with settings, and every choice has the chance to behave funny and produce complaints.

We for sure could design that functionality better, though I think much needs to be cleaned up before such design work will be effective. My plodding PRs and discussions and pestering for reviews this year are the early parts of that cleanup.

**Comment by rmrector** [comment]
jenkins build this please

File: xbmc/TextureCacheJob.cpp [hunk 1]
```diff
@@ -68,6 +68,36 @@ bool CTextureCacheJob::DoWork()
   return false;
 }
 
+namespace
+{
+// Most PVR images use "additional_info" to signify 'ownership' of basic images for easy
+// cache cleaning, rather than special generated images
+bool IsPVROwnedImage(std::string additional_info)
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:75]: please pass by const ref also for the other helper functions
```suggestion
bool IsPVROwnedImage(const std::string& additional_info)
```

File: xbmc/TextureCacheJob.cpp [hunk 2]
```diff
@@ -68,6 +68,36 @@ bool CTextureCacheJob::DoWork()
...
+// the GUI, so is not a special generated image
+bool IsControl(std::string additional_info)
+{
+  return additional_info == "flipped";
+}
+
+// special generated images and images served via HTTP should not be regularly checked for changes
+bool ShouldCheckForChanges(std::string additional_info, std::string url)
+{
+  bool isSpecialImage =
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:92]:
```suggestion
  const bool isSpecialImage =
```

File: xbmc/TextureCacheJob.cpp [hunk 3]
```diff
@@ -68,6 +68,36 @@ bool CTextureCacheJob::DoWork()
...
+      !additional_info.empty() && !IsControl(additional_info) && !IsPVROwnedImage(additional_info);
+  if (isSpecialImage)
+    return false;
+  bool isHTTP = StringUtils::StartsWith(url, "http://") || StringUtils::StartsWith(url, "https://");
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:96]:
```suggestion
  const bool isHTTP = StringUtils::StartsWith(url, "http://") || StringUtils::StartsWith(url, "https://");
```

File: xbmc/TextureCacheJob.cpp [hunk 4]
```diff
@@ -76,14 +106,17 @@ bool CTextureCacheJob::CacheTexture(std::unique_ptr<CTexture>* out_texture)
...
-    return false;
-  else if (m_details.hash == m_oldHash)
-    return true;
+  if (m_details.updateable)
+  {
+    // generate the hash
+    m_details.hash = GetImageHash(image);
+    if (m_details.hash.empty())
+      return false;
+    else if (m_details.hash == m_oldHash)
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:117]: `else` not needed (but doesn‘t harm and was there before)

File: xbmc/TextureCacheJob.cpp (hunk 1 above)
Comment (rmrector) [reply xbmc/TextureCacheJob.cpp]: Thanks, all updated.

**Comment by howie-f** [comment]
i'll give this a runtime go, if i understand correctly it restores old behavior - so i don't expect fallout.

> A common complaint is hard drives always spinning up for the hash check of images while scrolling through the library. 

mostly using local media so i doubt i can notice improvement with my setup

File: xbmc/TextureCacheJob.cpp [hunk 5]
```diff
@@ -68,6 +68,37 @@ bool CTextureCacheJob::DoWork()
...
+// special generated images and images served via HTTP should not be regularly checked for changes
+bool ShouldCheckForChanges(const std::string& additional_info, const std::string& url)
+{
+  const bool isSpecialImage =
+      !additional_info.empty() && !IsControl(additional_info) && !IsPVROwnedImage(additional_info);
+  if (isSpecialImage)
+    return false;
+  const bool isHTTP =
+      StringUtils::StartsWith(url, "http://") || StringUtils::StartsWith(url, "https://");
+  return !isHTTP;
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:98]:
```cpp
  const bool isSpecialImage =
      !additional_info.empty() && !IsControl(additional_info) && !IsPVROwnedImage(additional_info);
  if (isSpecialImage)
    return false;
    
  const bool isHTTP =
      StringUtils::StartsWith(url, "http://") || StringUtils::StartsWith(url, "https://");
  return !isHTTP;
```

File: xbmc/TextureCacheJob.cpp [hunk 6]
```diff
@@ -76,14 +107,17 @@ bool CTextureCacheJob::CacheTexture(std::unique_ptr<CTexture>* out_texture)
...
-    return true;
+  if (m_details.updateable)
+  {
+    // generate the hash
+    m_details.hash = GetImageHash(image);
+    if (m_details.hash.empty())
+      return false;
+    if (m_details.hash == m_oldHash)
+      return true;
+  }
```
Comment (howie-f) [inline xbmc/TextureCacheJob.cpp:120]:
```cpp
  if (m_details.updateable)
  {
    // generate the hash
    m_details.hash = GetImageHash(image);
    if (m_details.hash.empty())
      return false;
      
    if (m_details.hash == m_oldHash)
      return true;
  }
```

**Review by howie-f — APPROVED** [review]
maybe add a blank line here and there for readability. otherwise this looks good

**Comment by rmrector** [comment]
jenkins build this please

**Pull request merged**